drivers/i2c/i2c_nrfx_twi_rtio.c

```diff
@@ -0,0 +1,365 @@
+/*
+ * Copyright (c) 2018, Nordic Semiconductor ASA
+ * Copyright (c) 2024, Croxel Inc
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/drivers/i2c/rtio.h>
+#include <zephyr/dt-bindings/i2c/i2c.h>
+#include <zephyr/pm/device.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <soc.h>
+#include <nrfx_twi.h>
+
+#include <zephyr/logging/log.h>
+#include <zephyr/irq.h>
+LOG_MODULE_REGISTER(i2c_nrfx_twi, CONFIG_I2C_LOG_LEVEL);
+
+#if CONFIG_I2C_NRFX_TRANSFER_TIMEOUT
+#define I2C_TRANSFER_TIMEOUT_MSEC K_MSEC(CONFIG_I2C_NRFX_TRANSFER_TIMEOUT)
+#else
+#define I2C_TRANSFER_TIMEOUT_MSEC K_FOREVER
+#endif
+
+struct i2c_nrfx_twi_data {
+	struct i2c_rtio *ctx;
+	struct k_sem transfer_sync;
+	struct k_sem completion_sync;
+	volatile nrfx_err_t res;
+	uint32_t dev_config;
+};
+
+struct i2c_nrfx_twi_config {
+	nrfx_twi_t twi;
+	nrfx_twi_config_t config;
+	const struct pinctrl_dev_config *pcfg;
+};
+
+static int i2c_nrfx_twi_recover_bus(const struct device *dev);
+static void i2c_nrfx_twi_complete(const struct device *dev, int status);
+
+static bool i2c_nrfx_twi_msg_start(const struct device *dev, uint8_t flags,
+				   uint8_t *buf, size_t buf_len, uint16_t i2c_addr)
+{
+	const struct i2c_nrfx_twi_config *config = dev->config;
+	struct i2c_nrfx_twi_data *const dev_data = dev->data;
+	struct i2c_rtio *ctx = dev_data->ctx;
+	int ret = 0;
+
+	/** Enabling while already enabled ends up in a failed assertion: skip it. */
+	if (k_sem_take(&dev_data->completion_sync, K_NO_WAIT) == 0) {
+		nrfx_twi_enable(&config->twi);
+	}
+
+	if (flags & I2C_MSG_ADDR_10_BITS) {
+		LOG_ERR("10-bit I2C Addr devices not supported");
+		return i2c_rtio_complete(ctx, -ENOTSUP);
+	}
+
+	nrfx_twi_xfer_desc_t cur_xfer = {
+		.p_primary_buf = buf,
+		.primary_length = buf_len,
+		.address = i2c_addr,
+		.type = (flags & I2C_MSG_READ) ?
+			 NRFX_TWI_XFER_RX : NRFX_TWI_XFER_TX,
+	};
+	uint32_t xfer_flags = 0;
+	nrfx_err_t res;
+
+	/* In case the STOP condition is not supposed to appear after
+	 * the current message, check what is requested further:
+	 */
+	if (!(flags & I2C_MSG_STOP)) {
+		/* - if the transfer consists of more messages
+		 *   finish the transfer without generating the STOP
+		 *   condition, unless the current message is an RX
+		 *   request, for which such feature is not supported
+		 */
+		if (flags & I2C_MSG_READ) {
+			ret = -ENOTSUP;
+		} else {
+			xfer_flags |= NRFX_TWI_FLAG_TX_NO_STOP;
+		}
+	}
+
+	if (!ret) {
+		res = nrfx_twi_xfer(&config->twi, &cur_xfer, xfer_flags);
```

Comment:
Is there no interrupt to catch on completion of a transfer?

Reply:
`res` here just signals if the request started successfully. `event_handler()` triggers when the transfer completes (either successfully or in an error).

Does that make sense?